src/pyodide/internal/python.js

```diff
@@ -440,13 +433,22 @@ function encodeSnapshot(heap, dsoJSON) {
 /**
  * Decode heap and dsoJSON from the memory snapshot artifact we downloaded
  */
-function decodeSnapshot(memorySnapshot) {
-  const uint32View = new Uint32Array(memorySnapshot);
-  const snapshotOffset = uint32View[0];
-  const jsonLength = uint32View[1];
-  const jsonView = new Uint8Array(memorySnapshot, 8, jsonLength);
-  DSO_METADATA = JSON.parse(new TextDecoder().decode(jsonView));
-  MEMORY = new Uint8Array(memorySnapshot, snapshotOffset);
+function decodeSnapshot() {
+  const buf = new Uint32Array(2);
+  MEMORY_SNAPSHOT_READER.readMemorySnapshot(0, buf);
+  const snapshotOffset = buf[0];
+  const snapshotSize = MEMORY_SNAPSHOT_READER.getMemorySnapshotSize() - snapshotOffset;
+  const jsonLength = buf[1];
+  const jsonBuf = new Uint8Array(jsonLength);
+  MEMORY_SNAPSHOT_READER.readMemorySnapshot(8, jsonBuf);
+  DSO_METADATA = JSON.parse(new TextDecoder().decode(jsonBuf));
+  READ_MEMORY = function(Module) {
+    // resize linear memory to fit our snapshot.
+    Module.growMemory(snapshotSize);
```

Comment:
This allocates `snapshotSize` of memory, right? So we'll have 2*snapshotSize of memory allocated here (though not all of it will be attributed to the Worker I guess)

Is there a way we could avoid that? Could we directly assign the snapshot to `Module.HEAP8`?

Reply:
It appears that `INITIAL_MEMORY` is an Emscripten runtime setting that controls the size of the memory at startup. If we set this we can avoid the call to `growMemory` which is going to save some work.